Store points into file

jschleicher opened the pull request:
Closes #24

Todo: More module tests?

jschleicher committed:
drop unused coverage arg

jschleicher committed:
script to acquire coverage for unittests

to generate the report manually while not supported in catkin

jschleicher committed:
Added ros_message_serializer.py a tool to serialize into python formats which can be imported into other scripts

added basic test for serializer tool

jschleicher committed:
Improve Serializer Testcase with Timestamp

jschleicher committed:
convert rostest to unittest

and reduce code duplication

Hardthof commented:
The test case has 100% coverage for normal use. The edge cases (e.g. empty message, ...) should still be tested.

DamKoVosh commented on pilz_teleoperation/src/pilz_teleoperation/__init__.py:
```diff
@@ -18,3 +18,4 @@
 from .teleoperation_window import TeleoperationWindow
 from .curses_key_input import CursesKeyInput
 from .terminal_text_window import TerminalTextWindow
+from .ros_message_serializer import RosMessageSerializer
```
```suggestion
from .ros_message_serializer import RosMessageSerializer

```

DamKoVosh commented on pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py:
```diff
@@ -0,0 +1,53 @@
+import genpy
+import rospy
+
+class RosMessageSerializer:
```
allows for better inheritance
```suggestion
class RosMessageSerializer(object):
```

DamKoVosh commented on pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py:
```diff
@@ -0,0 +1,53 @@
+        """
+        Writes a file that can be used with 'import' to restore
+        :param ros_messages: iterable of messages to be serialized.
+                tuple contains the variable name used for serialization and a ROS message object
```
why using tuple and not dict since its supposed to be a key, value pair and dict is exactly that.

DamKoVosh commented on pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py:
```diff
@@ -0,0 +1,53 @@
+                serialized_message += "{} = {}\n".format(varname, self.convert_ros_message_to_python(msg))
+            f.write("\n".join("from {} import {}".format(module, typename) for typename, module in self._module_imports.items()))
+            f.write("\n\n")
+            f.write(serialized_message)
```
```suggestion
            f.write(serialized_message)
            f.write("\n")
```

DamKoVosh commented on pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py:
```diff
@@ -0,0 +1,53 @@
+        field_values = []  # ordered list with string-converted attribute values
+        for field_name, field_type in self._get_message_fields(message):
+            val = self.convert_ros_message_to_python(getattr(message, field_name), indentation + 2)
+            field_values.append((field_name,val))
```
again why not using dict?

Hardthof replied:
to maintain the order of the values in generated code.
Would you prefer an ordered dict?

DamKoVosh replied:
not sure if the order is relevant, if it is, ordered dict would be the way to go yes.

DamKoVosh commented on pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py:
```diff
@@ -0,0 +1,53 @@
+        
+        result = ',\n'.join(" "*(indentation+2) + '{} = {}'.format(key,value) for key, value in field_values)
+        if type(message).__name__ not in self._module_imports.keys():
+            self._module_imports[type(message).__name__] = type(message).__module__
```
As i see it, the if is useless, i doubt checking if its in there is much faster, than just putting it in no matter what

DamKoVosh commented on pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py:
```diff
@@ -0,0 +1,53 @@
+
+        return "{}(\n{}\n{})".format(type(message).__name__, result, " " * indentation)
+
+    def _get_message_fields(self, message):
```
can be static

DamKoVosh commented on pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py:
```diff
@@ -0,0 +1,53 @@
+        field_values = []  # ordered list with string-converted attribute values
+        for field_name, field_type in self._get_message_fields(message):
+            val = self.convert_ros_message_to_python(getattr(message, field_name), indentation + 2)
+            field_values.append((field_name,val))
```
maybe extract something in separate methods, and thus get rid of the comments

Hardthof replied:
aren't comments a good thing?

DamKoVosh replied:
Not necessarily => See Clean Code (Robert C. Martin)
Better if you explain yourself in code, and that can easily be done by extracting some code and use the name of a method to explain what you are doing.

DamKoVosh replied:
e.g.
`field_values = get_ordered_string_converted_attributes(message, indentation)`

DamKoVosh reviewed (changes requested):
Please also rebase to see if the testsetup still works

tobi-v commented:
> Please also rebase to see if the testsetup still works

This is already rebased against melodic_devel

Hardthof committed:
Fix formating in ros_message_serializer

Hardthof committed:
Incorporate Review Comments

Hardthof committed:
convert to pytest and add more data

to test also empty message, None type and multiple poses

Hardthof committed:
lint again

jschleicher committed:
refactor into OrderedDict

and split functionality into another helper function
to incorporate review comments

jschleicher committed:
add license headers

DamKoVosh commented on pilz_teleoperation/test/unittests/tst_ros_message_serializer.py:
```diff
@@ -0,0 +1,74 @@
+import sys
+import rospy
+import pytest
+from pilz_teleoperation.ros_message_serializer import RosMessageSerializer
```
its in __init__.py so module does not have to be named
```suggestion
from pilz_teleoperation import RosMessageSerializer
```

DamKoVosh commented on pilz_teleoperation/test/unittests/tst_ros_message_serializer.py:
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python
```
Filename should be test_... so pytest will be found (e.g. pycharm does not recognize them otherwise)
see: [Conventions for Python test discovery](https://docs.pytest.org/en/latest/goodpractices.html)

DamKoVosh commented on pilz_teleoperation/test/unittests/tst_ros_message_serializer.py:
```diff
@@ -0,0 +1,74 @@
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
```
not used
```suggestion
```

DamKoVosh commented on pilz_teleoperation/test/unittests/tst_ros_message_serializer.py:
```diff
@@ -0,0 +1,74 @@
+    """
+
+    # save Pose with unique name
+    module_name = str(tmpdir).split('/')[-1]
```
```suggestion
    module_name = tmpdir.basename
```

DamKoVosh replied:
not sure, why the unique modulname is necessary, since its a unique tmpdir already

test could be reduced to:
```
file_path = str(tmpdir / "test.py")
RosMessageSerializer().write_messages_to_file(test_input, file_path)
read_back = imp.load_source("test", file_path)

for k ...
```

Hardthof replied:
imp is deprecated for Python3?

DamKoVosh replied:
jes there we would have to use importlib. Don't know if the syntax would stay the same. If not we should find an alternative or leave it as it is.

DamKoVosh commented on pilz_teleoperation/test/unittests/tst_ros_message_serializer.py:
```diff
@@ -0,0 +1,74 @@
+    # load module
+    sys.path.insert(0, str(tmpdir))
+    readback = __import__(module_name)
+    sys.path.pop(0)
```
could be replaced with imp.load_source(module_name, file_path)

DamKoVosh replied:
just stumbled across [monkeypatching](https://docs.pytest.org/en/latest/monkeypatch.html),
maybe you can add the tmpdir to syspath with that `monkeypatch.syspath_prepend(path)`, gues that depends if you can get access to the tmpdir up there already.

DamKoVosh commented on pilz_teleoperation/src/pilz_teleoperation/ros_message_serializer.py:
```diff
@@ -0,0 +1,81 @@
+        field_values = self._message_to_stringdict(message, indentation)
+        result = ',\n'.join(
+            '{}{} = {}'.format(" " * (indentation + 2), key, value) for key, value in field_values.items())
+        self._module_imports[type(message).__name__] = type(message).__module__
```
type(message).__module__ results in imports like:
```
from geometry_msgs.msg._PoseStamped import PoseStamped
from geometry_msgs.msg._Point import Point
from geometry_msgs.msg._Pose import Pose
from std_msgs.msg._Header import Header
```

so it is using the private modul names, not the finally imported ones. It works, but you may want to think about resolving that, to prevent warnings when using this file

Hardthof replied:
currently okay as it is.

Hardthof committed:
Update pilz_teleoperation/test/unittests/tst_ros_message_serializer.py

Co-Authored-By: r.feistenauer <richard.feistenauer@gmx.de>

Hardthof committed:
add suggested change monkeypatch for temp path loading

Hardthof committed:
renamed tst files to test

Hardthof committed:
rename fixup

Hardthof committed:
Merge branch 'melodic-devel' into feature/storePoints
